### PR TITLE
Added an Ogg Vorbis decoder to mp3fs

### PR DESCRIPTION
**Pull request by quarterpounder**
Hello, 
in order to get a unified view of my flac and ogg music collection, I've added an Ogg Vorbis decoder class and integrated it into the build system. Please feel free to comment about it and, at your choice, incorporate it into your repo.

Kind regards,
Thomas Schwarzenberger

### CODE REVIEW

File: configure.ac [hunk 1]
```diff
@@ -17,7 +17,7 @@ AC_CHECK_SIZEOF(int)
 AS_IF([test "$ac_cv_sizeof_int" -lt 4],
     AC_MSG_ERROR([Integer type 'int' must be at least 4 bytes in size.]))
 
-# FLAC support checks
+# FLAC decoding support checks
```
Comment (khenriks) [inline configure.ac]: This should still just say FLAC support checks. The same library provides both encoding and decoding, even though we only use it for decoding.

File: configure.ac [hunk 2]
```diff
@@ -29,7 +29,19 @@ AS_IF([test "x$with_flac" != xno ],
 
 AM_CONDITIONAL([HAVE_FLAC], [test "x$with_flac" != xno])
 
-# MP3 support checks
+# Ogg Vorbis decoding support checks
```
Comment (khenriks) [inline configure.ac]: Ditto, this shouldn't reference decoding specifically.

File: configure.ac [hunk 3]
```diff
@@ -29,7 +29,19 @@ AS_IF([test "x$with_flac" != xno ],
...
+        [disable support for Ogg Vorbis decoding])],
+    [], [with_vorbisdec=yes])
+
+AS_IF([test "x$with_vorbisdec" != xno ],
+    [PKG_CHECK_MODULES([vorbisfile], [vorbisfile >= 1.3.0],
+        [AC_DEFINE([HAVE_VORBISDEC], [1], [Use vorbisfile library.])])])
+
+AM_CONDITIONAL([HAVE_VORBISDEC], [test "x$with_vorbisdec" != xno])
+
+# MP3 encoding support checks
```
Comment (khenriks) [inline configure.ac]: Ditto

File: mp3fs.1.txt [hunk 4]
```diff
@@ -1,26 +1,26 @@
 MP3FS(1)
 ========
-:doctype:	   	manpage
+:doctype:	manpage
```
Comment (khenriks) [inline mp3fs.1.txt]: Please don't change the indentation.

File: mp3fs.1.txt [hunk 5]
```diff
@@ -1,26 +1,26 @@
...
 NAME
 ----
-mp3fs - mounts and transcodes FLACs to MP3s on the fly
+mp3fs - mounts and transcodes FLACs and OGGs to MP3s on the fly
 
 
 SYNOPSIS
 --------
-*mp3fs* ['OPTION']... 'FLACDIR' 'MP3DIR'
+*mp3fs* ['OPTION']... 'FLACDIR|OGGDIR' 'MP3DIR'
```
Comment (khenriks) [inline mp3fs.1.txt]: This should probably be `IN_DIR` or something equally generic now.

File: src/coders.cc [hunk 6]
```diff
@@ -46,6 +49,10 @@ Decoder* Decoder::CreateDecoder(std::string file_type) {
 #ifdef HAVE_FLAC
     if (file_type == "flac") return new FlacDecoder();
 #endif
+#ifdef HAVE_VORBISDEC
+    if (file_type == "ogg" || file_type == "oga")
+		return new VorbisDecoder();
```
Comment (khenriks) [inline src/coders.cc]: Can you use braces here since the if statement is multi-line?

**Comment by khenriks** [comment]
Sorry for the wait. I'm catching up now on things that I've neglected. I appreciate your contribution.

The indentation and line length don't match the conventions used elsewhere in the code, and I'd like that to be more consistent. I'd also appreciate having tests added for this functionality. Are you interested in working further on this to respond to comments, or is this more of a code contribution that I can modify as I want to integrate it?

**Comment by quarterpounder** [comment]
Hello, thanks for your appreciation! Please feel free to streamline these changes into the way you structure your code. The Ogg Vorbis funtionality is working perfectly for me, but currently, it's hard for me to find spare time besides my profession.

I should be able to work on your comments as long as this is not too time critical. Test coverage possibly is a bigger problem.

File: INSTALL.md [hunk 7]
```diff
@@ -12,6 +12,7 @@ mp3fs requires the following libraries:
 
 * fuse (>= 2.6.0)
 * flac++ (>= 1.1.4)
+* libvorbisfile3
```
Comment (khenriks) [inline INSTALL.md]: This can just be libvorbis to match the common name for the library.

File: README.md [hunk 8]
```diff
@@ -71,6 +71,7 @@ mp3fs is written in a mixture of C and C++ and uses the following libraries:
 
 * [FUSE](http://fuse.sourceforge.net/)
 * [FLAC](http://flac.sourceforge.net/)
+* [libvorbisfile3] (http://www.xiph.org/vorbis/)
```
Comment (khenriks) [inline README.md]: Same comment as above.

File: mp3fs.1.txt [hunk 9]
```diff
@@ -1,26 +1,26 @@
 MP3FS(1)
 ========
-:doctype:	   	manpage
-:man source:   	mp3fs
+:doctype:       manpage
+:man source:    mp3fs
```
Comment (khenriks) [inline mp3fs.1.txt]: Please leave these with their previous indentation.

File: src/vorbis_decoder.cc [hunk 10]
```diff
@@ -0,0 +1,287 @@
...
+     * Use the Replay Gain tag to set volume scaling. The appropriate
+     * value for dbgain is set in the above if statements according to
+     * the value of gainmode. Obey the gainref option here.
+     */
+    if (dbgain != INVALID_DB_GAIN) {
+        encoder->set_gain_db(params.gainref - filegainref + dbgain);
+    }
+
+    /*
+     * TODO: decode METADATA_BLOCK_PICTURE for use with
```
Comment (khenriks) [inline src/vorbis_decoder.cc]: Can you do this TODO? It should be pretty straightforward if the libvorbis library has a way to access the picture structure.

**Comment by khenriks** [comment]
I've just added a couple comments to resolve, and test coverage can come later if necessary (as long as the existing tests still pass with the new code).

File: src/vorbis_decoder.cc (hunk 10 above)
Comment (quarterpounder) [inline src/vorbis_decoder.cc]: Should be possible, maybe soon.

**Comment by quarterpounder** [comment]
Ogg Vorbis picture handling is working for valid METADATA_BLOCK_PICTURE.

**Comment by quarterpounder** [comment]
Hi there, maybe you'd like to have a look at the progress here. By now, the Ogg Vorbis part should be complete, inclusive picture handling.

**Comment by khenriks** [comment]
Thanks so much! Sorry for the delay, and I appreciate your contribution.

**Comment by bluemoehre** [comment]
Thanks a lot! I will update as soon as possible =)

**Comment by intrigeri** [comment]
Any chance a new release is published? I'd love to use this feature once this reaches Debian :)

**Pull request merged**